### PR TITLE
[Feature] [kubectl-plugin] Expose setting `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` in ray job submit

### LINKED ISSUE #3560: [Feature] [kubectl-plugin] Expose setting `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` in ray job submit
### Search before asking

- [x] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

As per title, currently `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` aren't being exposed in `kubectl ray job submit`. Ideally these two can also set as flags when using the kubectl plugin: 

```bash
kubectl ray job submit --shutdown-after-job-finishes --ttl-seconds-after-finished 10 ...
```

### Use case

I want ray jobs created by users to be automatically cleaned up, rather than be left around. This can be set if the user were to create and manage their own CRD yaml files, but not when using the plugin CLI. 

In particular, after a ray job is submitted and completed, and another ray job with the same name is submitted, this returns an error since there's no cleanup of the previous RayJob CRD.

### Related issues

_No response_

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by CheyuWu**
## Why are these changes needed?
Currently shutdownAfterJobFinishes and ttlSecondsAfterFinished aren't being exposed in kubectl ray job submit, although they are available in the [CRD](https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/rayjob-quick-start.html#rayjob-configuration).

### Ray Operator

- Add support for `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` in RayJob spec.

- Implement validation logic: if `shutdownAfterJobFinishes` is False, then `ttlSecondsAfterFinished` must be 0 or unset.

- Add corresponding unit tests.

### kubectl-plugin
- Add `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` flags to the rayjob submit command.

- Implement validation logic: if `shutdownAfterJobFinishes` is False, then `ttlSecondsAfterFinished` must be 0 or unset.

- Add corresponding unit tests.

## Related issue number
Closes #3560 

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [ ] Manual tests
  - [ ] This PR is not tested :(

### CODE REVIEW

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 1]
```diff
@@ -40,40 +40,42 @@ const (
...
+	namespace                string
+	entryPointMemory         int
+	entryPointGPU            float32
+	workerReplicas           int32
+	entryPointCPU            float32
+	noWait                   bool
+	dryRun                   bool
+	verbose                  bool
+	shutdownAfterJobFinishes bool
+	ttlSecondsAfterFinished  int32
```
Comment (CheyuWu) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:78]: Add `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished`

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 2]
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 			}
 			options.runtimeEnvJson = string(runtimeJson)
 		}
+
+		if !options.RayJob.Spec.ShutdownAfterJobFinishes && options.RayJob.Spec.TTLSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
+
```
Comment (CheyuWu) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:265]: Implement validation logic when `--filename` is provided
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 3]
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
...
 	} else if strings.TrimSpace(options.rayjobName) == "" {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
```
Comment (CheyuWu) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:271]: Implement validation logic when `--filename` is not provided
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

File: kubectl-plugin/pkg/cmd/job/job_submit_test.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ func TestRayJobSubmitComplete(t *testing.T) {
 	assert.Equal(t, "fake/path/to/env/yaml", fakeSubmitJobOptions.runtimeEnv)
 }
 
-func TestRayJobSubmitValidate(t *testing.T) {
+func TestRayJobSubmitWithYamlValidate(t *testing.T) {
```
Comment (CheyuWu) [inline kubectl-plugin/pkg/cmd/job/job_submit_test.go:33]: Split TestRayJobSubmitValidate into two test cases: one for YAML-based submission and one for non-YAML-based submission.

File: kubectl-plugin/pkg/cmd/job/job_submit_test.go [hunk 5]
```diff
@@ -99,6 +122,52 @@ spec:
 	}
 }
 
+func TestRayJobSubmitWithoutYamlValidate(t *testing.T) {
```
Comment (CheyuWu) [inline kubectl-plugin/pkg/cmd/job/job_submit_test.go:125]: Split TestRayJobSubmitValidate into two test cases: one for YAML-based submission and one for non-YAML-based submission.

File: ray-operator/controllers/ray/utils/validation.go [hunk 6]
```diff
@@ -144,6 +144,10 @@ func ValidateRayJobSpec(rayJob *rayv1.RayJob) error {
 		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false is not allowed to be suspended")
 	}
 
+	if !rayJob.Spec.ShutdownAfterJobFinishes && rayJob.Spec.TTLSecondsAfterFinished != 0 {
+		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished is non-zero")
+	}
+
```
Comment (CheyuWu) [inline ray-operator/controllers/ray/utils/validation.go:150]: Implement validation logic in ray operator
> if shutdownAfterJobFinishes is False, then ttlSecondsAfterFinished must be 0 or unset.

**Comment by CheyuWu** [comment]
Hi @MortalHappiness and @kevin85421  PTAL

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 7]
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
 			}
 			options.runtimeEnvJson = string(runtimeJson)
 		}
+
+		if !options.RayJob.Spec.ShutdownAfterJobFinishes && options.RayJob.Spec.TTLSecondsAfterFinished != 0 {
```
Comment (davidxia) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:262]: unrelated to this PR, but when I was testing this change, I ran `kubectl ray job submit -f ray-job.sample.yaml --working-dir . --shutdown-after-job-finishes --ttl-seconds-after-finished 600 -- python /home/ray/samples/sample_code.py` and wondered why the new flags weren't used. I had to look at this source to understand `-f/--filename` makes the command ignore many of these other flags.

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 8]
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
...
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
+
 	} else if strings.TrimSpace(options.rayjobName) == "" {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
```
Comment (davidxia) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:270]:
```suggestion
			return fmt.Errorf("--ttl-seconds-after-finished is only supported when --shutdown-after-job-finishes is set to true")
```

File: ray-operator/controllers/ray/utils/validation.go [hunk 9]
```diff
@@ -144,6 +144,10 @@ func ValidateRayJobSpec(rayJob *rayv1.RayJob) error {
 		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false is not allowed to be suspended")
 	}
 
+	if !rayJob.Spec.ShutdownAfterJobFinishes && rayJob.Spec.TTLSecondsAfterFinished != 0 {
+		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished is non-zero")
```
Comment (davidxia) [inline ray-operator/controllers/ray/utils/validation.go:148]:
```suggestion
		return fmt.Errorf("a RayJob with shutdownAfterJobFinishes set to false cannot have TTLSecondsAfterFinished")
```

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 10]
```diff
@@ -254,8 +258,17 @@ func (options *SubmitJobOptions) Validate() error {
...
 		}
+
+		if !options.RayJob.Spec.ShutdownAfterJobFinishes && options.RayJob.Spec.TTLSecondsAfterFinished != 0 {
+			return fmt.Errorf("ttl-seconds-after-finished is only supported when shutdown-after-job-finishes is set to true")
+		}
+
 	} else if strings.TrimSpace(options.rayjobName) == "" {
 		return fmt.Errorf("Must set either yaml file (--filename) or set Ray job name (--name)")
+	} else if options.fileName == "" {
+		if !options.shutdownAfterJobFinishes && options.ttlSecondsAfterFinished != 0 {
```
Comment (davidxia) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:269]: also check ttlSecondsAfterFinished ≥ 0 in both cases? Right now this command will create a RayJob with ttlSecondsAfterFinished: -10

```
kubectl ray --context gke_kubeflow-platform_europe-west4-b_ml-compute-1 -n hyperkube job submit --name dxia-test --shutdown-after-job-finishes --ttl-seconds-after-finished -10 --working-dir . --dry-run  -- python /home/ray/samples/sample_code.py | grep -B 1 ttlSeconds
  submissionMode: InteractiveMode
  ttlSecondsAfterFinished: -10
```

File: apiserver/pkg/model/converter_test.go [hunk 11]
```diff
@@ -303,9 +303,10 @@ var JobNewClusterTest = rayv1api.RayJob{
 		Metadata: map[string]string{
 			"job_submission_id": "123",
 		},
-		RuntimeEnvYAML:          "mytest yaml",
-		TTLSecondsAfterFinished: secondsValue,
-		RayClusterSpec:          &ClusterSpecTest.Spec,
+		RuntimeEnvYAML:           "mytest yaml",
+		ShutdownAfterJobFinishes: true,
```
Comment (MortalHappiness) [inline apiserver/pkg/model/converter_test.go:307]: Why change apiserver tests?

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 12]
```diff
@@ -171,6 +173,8 @@ func NewJobSubmitCommand(cmdFactory cmdutil.Factory, streams genericclioptions.I
 	cmd.Flags().StringVar(&options.workerGPU, "worker-gpu", "0", "number of GPUs in each worker group replica")
 	cmd.Flags().BoolVar(&options.dryRun, "dry-run", false, "print the generated YAML instead of creating the cluster. Only works when filename is not provided")
 	cmd.Flags().BoolVarP(&options.verbose, "verbose", "v", false, "Passing the '--verbose' flag to the 'ray job submit' command")
+	cmd.Flags().BoolVar(&options.shutdownAfterJobFinishes, "shutdown-after-job-finishes", false, "Shutdown the cluster after the job finishes")
+	cmd.Flags().Int32Var(&options.ttlSecondsAfterFinished, "ttl-seconds-after-finished", 0, "TTL seconds after finished. Only works when filename is not provided")
```
Comment (MortalHappiness) [inline kubectl-plugin/pkg/cmd/job/job_submit.go:177]: It still works when filename is provided right?

File: kubectl-plugin/pkg/cmd/job/job_submit.go (hunk 7 above)
Comment (MortalHappiness) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:262]: +1. I think we should take the filename as a base, and flags will override the values in the file.

I created an issue to track this https://github.com/ray-project/kuberay/issues/3639

Comment (davidxia) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:262]: The behavior of `kubectl ray create cluster` is to error with an actionable error message if declarative file and CLI flags are used together. We can consider doing that here too if we think it's not worth it to have override logic.

Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:262]: It is worth overriding the values in the file. I will change this.

File: kubectl-plugin/pkg/cmd/job/job_submit.go (hunk 12 above)
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:177]: Currently, I will make sure it work in next commit.

File: apiserver/pkg/model/converter_test.go (hunk 11 above)
Comment (CheyuWu) [reply apiserver/pkg/model/converter_test.go:307]: Because I have modified the ray operator that `ShutdownAfterJobFinishes` should be true; otherwise, `ttlSecondsAfterFinishedcannot`  cannot be set. 
[reference](https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/rayjob-quick-start.html#rayjob-configuration)

File: kubectl-plugin/pkg/cmd/job/job_submit.go (hunk 10 above)
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:269]: I will fix this, thx

File: apiserver/pkg/model/converter_test.go (hunk 11 above)
Comment (MortalHappiness) [reply apiserver/pkg/model/converter_test.go:307]: Seems unrelated to this PR? Could you do apiserver-related changes in a separate PR?

**Comment by CheyuWu** [comment]
Hi @davidxia, @MortalHappiness, PTAL
- Fixed the error message for better clarity.

- When using `--shutdown-after-job-finishes` or `--ttl-seconds-after-finished`, these flags will override the corresponding fields (`shutdownAfterJobFinishes` or `ttlSecondsAfterFinished`) in the file specified by `--filename`.

- Added validation to reject negative values for `ttlSecondsAfterFinished` or the `--ttl-seconds-after-finished` flag.

- Reverted changes related to the Ray Operator.

File: kubectl-plugin/pkg/cmd/job/job_submit_test.go [hunk 13]
```diff
@@ -71,6 +73,41 @@ spec:
...
+metadata:
+  name: rayjob-sample
+spec:
+  shutdownAfterJobFinishes: false
+  ttlSecondsAfterFinished: 10
+  submissionMode: 'InteractiveMode'`,
+			expectError: "ttlSecondsAfterFinished is only supported when shutdownAfterJobFinishes is set to true",
+		},
+		{
+			name: "shutdownAfterJobFinishes is false and ttlSecondsAfterFinished is not zero",
```
Comment (MortalHappiness) [inline kubectl-plugin/pkg/cmd/job/job_submit_test.go:89]:
```suggestion
			name: "shutdownAfterJobFinishes is true and ttlSecondsAfterFinished is not zero",
```

Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit_test.go:89]: Ops,

**Review by MortalHappiness — APPROVED** [review]

**Review by davidxia — APPROVED** [review]

**Review by kevin85421 — COMMENTED** [review]
Why do we need to expose `shutdownAfterJobFinishes` and `ttlSecondsAfterFinished` at the same time? We can expose `ttlSecondsAfterFinished` only and configure `shutdownAfterJobFinishes` to true when `ttlSecondsAfterFinished` is set.

also cc @MortalHappiness @davidxia

**Comment by davidxia** [comment]
ah that's true

**Comment by MortalHappiness** [comment]
SGTM. @CheyuWu Could you only expose `ttlSecondsAfterFinished`? Thanks.

**Comment by CheyuWu** [comment]
Hi @kevin85421 @MortalHappiness  @davidxia PTAL
I have updated it to expose only `ttlSecondsAfterFinished`.

## Manual Testing
> Set shutdownAfterJobFinishes to false and ttlSecondsAfterFinished to 2
> my_yaml.yaml
```yaml
apiVersion: ray.io/v1
kind: RayJob
metadata:
  name: rayjob-sample
spec:
  submissionMode: 'InteractiveMode'
  shutdownAfterJobFinishes: false 
  ttlSecondsAfterFinished: 2
  rayClusterSpec:
    rayVersion: '2.41.0'
    headGroupSpec:
      rayStartParams: {}
      template:
        spec:
          containers:
            - name: ray-head
              image: rayproject/ray:2.41.0
              ports:
                - containerPort: 6379
                  name: gcs-server
                - containerPort: 8265
                  name: dashboard
                - containerPort: 10001
                  name: client
              resources:
                limits:
                  cpu: "1"
                requests:
                  cpu: "200m"
    workerGroupSpecs:
      - replicas: 1
        minReplicas: 1
        maxReplicas: 5
        groupName: small-group
        rayStartParams: {}
        template:
          spec:
            containers:
              - name: ray-worker
                image: rayproject/ray:2.41.0
                resources:
                  limits:
                    cpu: "1"
                  requests:
                    cpu: "200m"

```

Only override ttlSecondsAfterFinished -> Success
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 10 -- python entrypoint-python-sample.py
```

set `--ttl-seconds-after-finished` negative value -> failure
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished -10 -- python entrypoint-python-sample.py
```
```console
Error: --ttl-seconds-after-finished must be greater than or equal to 0
```

Not override yaml config
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml  -- python entrypoint-python-sample.py
```
```console
Error: ttlSecondsAfterFinished is only supported when shutdownAfterJobFinishes is set to true
```

File: kubectl-plugin/pkg/cmd/job/job_submit.go [hunk 14]
```diff
@@ -40,42 +40,44 @@ const (
...
+	workerGPU                string
+	namespace                string
+	entryPointMemory         int
+	entryPointGPU            float32
+	workerReplicas           int32
+	entryPointCPU            float32
+	noWait                   bool
+	dryRun                   bool
+	verbose                  bool
+	shutdownAfterJobFinishes bool
```
[the remark this answers is not in the document]
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: Hi @MortalHappiness, If we remove `shutdownAfterJobFinishes`, it will cause an issue [here](https://github.com/ray-project/kuberay/pull/3627/files#diff-d7990c8ae1278889923e749905d9699559a9993a95a6bb7e7b62cca21968c3c7R144), where we won't be able to assign a value when generating the RayJob YAML.

If we try to determine `ShutdownAfterJobFinishes` based on the `ttlSecondsAfterFinished` field, there will be a problem when `ttlSecondsAfterFinished` is explicitly set to 0 (which is its default value).

Please let me know if I’ve misunderstood anything or missed any details.

[the remark this answers is not in the document]
Comment (MortalHappiness) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: Ok there are some follow-up questions:

* Did you try setting `ttlSecondsAfterFinished` to `0` manually in the CLI, and does it work?
  * If yes, add a test that explicitly sets it to 0 and verify that it works.
  * If not, we can probably set `SubmitJobOptions.ttlSecondsAfterFinished` to `-1` in `Validate` if the user doesn't provide a value. Then in `Run`, we only set `ShutdownAfterJobFinishes` to true when the value is non-negative.

[the remark this answers is not in the document]
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: I ran the commands
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit -f testdata/my_yaml.yaml --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 0 -- python entrypoint-python-sample.py
```
And it works
image

Also, 
```bash
$ cd kubectl-plugin/test/e2e
$ kubectl ray job submit --name rayjob-sample --runtime-env testdata/rayjob-submit-working-dir/runtime-env-sample.yaml --ttl-seconds-after-finished 0 -- python entrypoint-python-sample.py
```
And it works
image

[the remark this answers is not in the document]
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: I got an idea, we probably can set `ttl-seconds-after-finished` to -1 as default?

[the remark this answers is not in the document]
Comment (MortalHappiness) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: It's ok if both setting it to `0` and omitting the flag work as expected. You don't need to set it to `-1`. You can preserve the `shutdownAfterJobFinishes` field. But could you add e2e tests for these two corner cases? Thanks.

[the remark this answers is not in the document]
Comment (CheyuWu) [reply kubectl-plugin/pkg/cmd/job/job_submit.go:79]: OK, no problem

**Comment by CheyuWu** [comment]
Hi @MortalHappiness PTAL

File: kubectl-plugin/test/e2e/kubectl_ray_job_submit_test.go [hunk 15]
```diff
@@ -100,6 +100,68 @@ var _ = Describe("Calling ray plugin `job submit` command on Ray Job", func() {
...
+		)
+		output, err := cmd.CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
+		// Retrieve the Job ID from the output
+		cmdOutputJobID := extractRayJobID(string(output))
+
+		rayJob := getAndCheckRayJob(namespace, rayJobName, cmdOutputJobID, "SUCCEEDED", "Complete")
+		Expect(rayJob.Spec.TTLSecondsAfterFinished).To(Equal(int32(0)))
+		Expect(rayJob.Spec.ShutdownAfterJobFinishes).To(BeTrue())
+	})
```
Comment (MortalHappiness) [inline kubectl-plugin/test/e2e/kubectl_ray_job_submit_test.go:128]: Add another testcase for `--ttl-seconds-after-finished` is not used at all.

**Review by MortalHappiness — APPROVED** [review]

**Review by kevin85421 — APPROVED** [review]
